Add examples for NVRAM usage

tomoveu opened the pull request:
This PR is the second part of ZD11566 and adds two new examples:
* examples/nvram/store
* examples/nvram/read

Examples work without parameter encryption.

With parameter encryption, NV DefineSpace breaks execution (working on this issue)

1. Generate a TPM key (can be any key, RSA, ECC or AES symmetric)
```
pi@raspberrypi:~/wolfTPM $ sudo ./examples/keygen/keygen -rsa
TPM2.0 Key generation example
	Key Blob: keyblob.bin
	Algorithm: RSA
	Template: AIK
	Use Parameter Encryption: NULL
wolfSSL Entering wolfCrypt_Init
TPM2: Caps 0x00000000, Did 0x0000, Vid 0x0000, Rid 0x 0 
TPM2_ReadPublic Handle 0x81000200: pub 282, name 34, qualifiedName 34
Loading SRK: Storage 0x81000200 (282 bytes)
RSA AIK template
Creating new RSA key...
Created new key (pub 280, priv 222 bytes)
Wrote 840 bytes to keyblob.bin
wolfSSL Entering wolfCrypt_Cleanup
```

2. Store the key in the TPM's NVRAM and set password (the string gNvAuth from tpm_test.h)

```
pi@raspberrypi:~/wolfTPM $ sudo examples/nvram/store
wolfSSL Entering wolfCrypt_Init
TPM2: Caps 0x00000000, Did 0x0000, Vid 0x0000, Rid 0x 0 
Reading 840 bytes from keyblob.bin
TPM2_NV_DefineSpace: Auth 0x40000001, Idx 0x1800203, Attribs 0x33947654, Size 1024
Storing key at TPM NV index 0x1800203 with password protection
Public part = 616 bytes
TPM2_NV_Write: Auth 0x1800203, Idx 0x1800203, Offset 0, Size 616
Private part = 222 bytes
TPM2_NV_Write: Auth 0x1800203, Idx 0x1800203, Offset 616, Size 222
NV write succeeded
wolfSSL Entering wolfCrypt_Cleanup
```

3. Read out the key from NVRAM using the password (requires to specify size of the private and public part)

```
pi@raspberrypi:~/wolfTPM $ sudo examples/nvram/read 616 222
wolfSSL Entering wolfCrypt_Init
TPM2: Caps 0x00000000, Did 0x0000, Vid 0x0000, Rid 0x 0 
Trying to read 616 bytes of public key part from NV
TPM2_NV_Read: Auth 0x1800203, Idx 0x1800203, Offset 0, Size 616
Trying to read 222 bytes of private key part from NV
TPM2_NV_Read: Auth 0x1800203, Idx 0x1800203, Offset 616, Size 222
TPM2_NV_UndefineSpace: Auth 0x40000001, Idx 0x1800203
Extraction of key from NV RAm at index 0x1800203
wolfSSL Entering wolfCrypt_Cleanup
```

First part of Zd11566 is #143

tomoveu commented:
I found an issue with some TPM2 Commands who get their packet data overwritten by the HMAC field calculated during parameter encryption. This was happening because no space was reserved for the HMAC field (hmac.Size/auth.Size was 0 during AppendAuth), so packet->pos was pointing to an incorrect place in the packet buffer later during ParamEnc functions.

After figuring out the tricky situation, I found a nice place in SetAuthSession for the fix. Because this issue only impacts Parameter Encryption sessions at the moment. In future, it could have impact on HMAC sessions without parameter encryption.

tomoveu commented:
AUTH HMAC issue comes from the Handle Names used during authorization.  The reason is that we are extracting the names incorrectly.

To complete the fix and extract(compute) the correct names we might need to start managing information about TPM objects that the stack creates.

I pushed my draft code for the fix with comment explaining the TPM object info we miss.

tomoveu commented:
I just noticed that some of my new code was lost after rebase. I pushed it back as a new commit.

And this made me realize that **we might also need the original TPM2_GetName, but as TPM2_GetSessionName for calcCpHash().**

tomoveu commented:
I just noticed that some of my new code was lost after rebase. I pushed it back as a new commit.

dgarske commented on src/tpm2.c:
```diff
@@ -187,9 +188,16 @@ static int TPM2_CommandProcess(TPM2_CTX* ctx, TPM2_Packet* packet,
-            rc =  TPM2_GetName(ctx, info->inHandleCnt, 0, &name1);
-            rc |= TPM2_GetName(ctx, info->inHandleCnt, 1, &name2);
-            rc |= TPM2_GetName(ctx, info->inHandleCnt, 2, &name3);
+            UINT32 handleValue;
```
Please move variable declaration to top of function or put within a brace.

dgarske replied:
This is still an issue. Variables must be declared at top of function or brace.

tomoveu replied:
Indeed. Fixed

dgarske commented on src/tpm2_wrap.c:
```diff
@@ -2560,9 +2582,42 @@ int wolfTPM2_NVCreateAuth(WOLFTPM2_DEV* dev, WOLFTPM2_HANDLE* parent,
 
+    /* Compute NV Index name in case of parameter encryption */
+#ifndef WOLFTPM2_NO_WOLFCRYPT
+    rc = TPM2_GetHashType(in.publicInfo.nvPublic.nameAlg);
```
Please move this logic into its own function that takes in an nvPublic struct computes the hash digest (name for NV).

tomoveu replied:
Already move into TPM2_HashNvPublic

dgarske commented on src/tpm2.c:
```diff
@@ -5686,6 +5706,20 @@ void TPM2_PrintBin(const byte* buffer, word32 length)
+        return;
+
+    printf("authCmd:\n");
+    printf("sessionHandle=0x%7X\n", authCmd->sessionHandle);
```
Why not `0x%08X` or `0x%8X`?

tomoveu replied:
session Handles are 7 digits, other TPM objects are loaded into 8 digit handles

```
wolftpm/tpm2.h:    TPM_HT_PCR              = 0x00,
wolftpm/tpm2.h:    TPM_HT_NV_INDEX         = 0x01,
wolftpm/tpm2.h:    TPM_HT_HMAC_SESSION     = 0x02,
wolftpm/tpm2.h:    TPM_HT_LOADED_SESSION   = 0x02,
wolftpm/tpm2.h:    TPM_HT_POLICY_SESSION   = 0x03,
wolftpm/tpm2.h:    TPM_HT_ACTIVE_SESSION   = 0x03,

notice the leading zero, of course we could print that too to show it is a 7 digit session handle

wolftpm/tpm2.h:    TPM_HT_PERMANENT        = 0x40,
wolftpm/tpm2.h:    TPM_HT_TRANSIENT        = 0x80,
wolftpm/tpm2.h:    TPM_HT_PERSISTENT       = 0x81,

```

I will fix it.

tomoveu commented:
@dgarske  thanks, I pushed new changes

* Hashing over the whole infoPublic instead of just nvPublic
* added generic function for Hashing TPM2B_NV_PUBLIC

Strangely, the issue persists, my name computation differs from what the tpm expects.

tomoveu commented:
@dgarske  ready for review :)

Once done, I will rebase the changes to 2-3 commits for better history readability. 

And for GitHub issue traceability:
- Fixed TPM2_GetName computation
- Fixed HMAC auth issue due nvPublic changing with NV_WRITTEN flag between NVWrite commands, thus Name chagned and HMAC Auth failed 
- Fixed unneeded Auth slot clearing in wolfTPM2_NVDeleteAuth wrapper and added new helper

**Output with parameter encryption**

```
dimitartomov@Dimitars-MacBook-Pro wolfTPM % ./examples/nvram/store -aes
...
TPM2_StartAuthSession: handle 0x2000000, algorithm AES
TPM2_StartAuthSession: sessionHandle 0x2000000
Reading 840 bytes from keyblob.bin
TPM2_NV_DefineSpace: Auth 0x40000001, Idx 0x1800203, Attribs 0x33947654, Size 1024
Storing key at TPM NV index 0x1800203 with password protection
Public part = 616 bytes
TPM2_NV_Write: Auth 0x1800203, Idx 0x1800203, Offset 0, Size 616
Private part = 222 bytes
TPM2_NV_Write: Auth 0x1800203, Idx 0x1800203, Offset 616, Size 222
**NV write succeeded**
TPM2_FlushContext: Closed handle 0x2000000

dimitartomov@Dimitars-MacBook-Pro wolfTPM % ./examples/nvram/read 616 222 -aes
...
TPM2_StartAuthSession: handle 0x2000000, algorithm AES
TPM2_StartAuthSession: sessionHandle 0x2000000
Trying to read 616 bytes of public key part from NV
TPM2_NV_Read: Auth 0x1800203, Idx 0x1800203, Offset 0, Size 616
Trying to read 222 bytes of private key part from NV
TPM2_NV_Read: Auth 0x1800203, Idx 0x1800203, Offset 616, Size 222
TPM2_NV_UndefineSpace: Auth 0x40000001, Idx 0x1800203
Extraction of key from NVRAM at index 0x1800203 succeeded
TPM2_FlushContext: Closed handle 0x2000000
dimitartomov@Dimitars-MacBook-Pro wolfTPM % 

```

Output without parameter encryption
```
dimitartomov@Dimitars-MacBook-Pro wolfTPM % ./examples/nvram/store     
...
Reading 840 bytes from keyblob.bin
TPM2_NV_DefineSpace: Auth 0x40000001, Idx 0x1800203, Attribs 0x33947654, Size 1024
Storing key at TPM NV index 0x1800203 with password protection
Public part = 616 bytes
TPM2_NV_Write: Auth 0x1800203, Idx 0x1800203, Offset 0, Size 616
Private part = 222 bytes
TPM2_NV_Write: Auth 0x1800203, Idx 0x1800203, Offset 616, Size 222
NV write succeeded

dimitartomov@Dimitars-MacBook-Pro wolfTPM % ./examples/nvram/read 616 222     
...
Trying to read 616 bytes of public key part from NV
TPM2_NV_Read: Auth 0x1800203, Idx 0x1800203, Offset 0, Size 616
Trying to read 222 bytes of private key part from NV
TPM2_NV_Read: Auth 0x1800203, Idx 0x1800203, Offset 616, Size 222
TPM2_NV_UndefineSpace: Auth 0x40000001, Idx 0x1800203
Extraction of key from NVRAM at index 0x1800203 succeeded
dimitartomov@Dimitars-MacBook-Pro wolfTPM % 
```

dgarske commented on src/tpm2.c:
```diff
@@ -5278,10 +5287,28 @@ int TPM2_GetName(TPM2_CTX* ctx, int handleCnt, int idx, TPM2B_NAME* name)
 
     session = &ctx->session[idx];
+
+#if 0
```
Please remove dead code.

tomoveu replied:
Removed. Thanks!

dgarske commented on src/tpm2.c:
```diff
@@ -5648,6 +5675,67 @@ UINT16 TPM2_GetVendorID(void)
+
+    wc_HashFree(&hash, hashType);
+
+    return TPM_RC_SUCCESS;
```
Return code is not being checked or returned... Where is `rc` checked?

tomoveu replied:
this is changed to return rc; now

and in the case when wolfcrypt is not enabled(param encryption is not enabled) I return TPM_RC_SUCCESS, because the computation will not be used without param encryption.

tomoveu replied:
the RC result is returned from the wolfTPM2_wrappers to the caller

this is why the RC is not checked, I will add comment so it is clear

dgarske commented on src/tpm2_wrap.c:
```diff
@@ -26,6 +26,11 @@
+#ifndef WOLFTPM2_NO_WOLFCRYPT
+/* Required in wolfTPM2_CreateAuth() for name computation of NV handles */
+#include <wolfssl/wolfcrypt/hash.h>
+#endif
```
I believe this section is no longer required because its in tpm2.c now.

tomoveu replied:
Indeed. Good catch

dgarske commented on src/tpm2_wrap.c:
```diff
@@ -369,6 +374,20 @@ int wolfTPM2_GetCapabilities(WOLFTPM2_DEV* dev, WOLFTPM2_CAPS* cap)
+{
+    TPM2_AUTH_SESSION* session;
+
+    if (dev == NULL || index >= MAX_SESSION_NUM) {
```
What about a negative index? I see other cases where this could happen. Please correct.

tomoveu replied:
I took this rule from SetAuth, so I will apply a fix there as well.

dgarske commented on src/tpm2_wrap.c:
```diff
@@ -2524,6 +2570,7 @@ int wolfTPM2_NVCreateAuth(WOLFTPM2_DEV* dev, WOLFTPM2_HANDLE* parent,
 {
     int rc;
     NV_DefineSpace_In in;
+    TPM2B_NAME name;
```
You can use `nv->handle.name` directly. A separate TPM2B_NAME is not required.

tomoveu replied:
Indeed. I was already doing it in WriteAuth and ReadAuth. Good catch. Thank you

dgarske commented on src/tpm2_wrap.c:
```diff
@@ -2605,6 +2661,27 @@ int wolfTPM2_NVWriteAuth(WOLFTPM2_DEV* dev, WOLFTPM2_NV* nv,
         wolfTPM2_SetAuthHandle(dev, 0, &nv->handle);
     }
 
+    XMEMSET((byte*)&inPublic, 0, sizeof(inPublic));
```
Why not use `wolfTPM2_NVReadPublic`?

tomoveu replied:
Wow. I've forgot we have wolfTPM2 wrapper for that. Changed the code to use it - Awesome :)

dgarske commented on src/tpm2_wrap.c:
```diff
@@ -2665,6 +2745,27 @@ int wolfTPM2_NVReadAuth(WOLFTPM2_DEV* dev, WOLFTPM2_NV* nv,
         wolfTPM2_SetAuthHandle(dev, 0, &nv->handle);
     }
 
+    XMEMSET((byte*)&inPublic, 0, sizeof(inPublic));
```
Why not use `wolfTPM2_NVReadPublic`?

tomoveu replied:
Likewise, awesome to use our own wrappers and make the code even better and more neat.

dgarske commented on src/tpm2.c:
```diff
@@ -5648,6 +5675,67 @@ UINT16 TPM2_GetVendorID(void)
+    wc_HashAlg hash;
+    enum wc_HashType hashType;
+    byte appending[sizeof(TPMS_NV_PUBLIC)];
+    TPM2B_DIGEST digest;
```
This is not guaranteed to be max digest + 2. Use TPM2B_NAME with .name. That is... `TPMT_HA`.

tomoveu replied:
Very good point - indeed TPM2B_DIGEST buffer can not fit digest+2, but only digest size.

I will use TPM2B_DATA , because this is the type with buffer of size TPMT_HA
```

typedef struct TPM2B_DATA {
    UINT16 size;
    BYTE buffer[sizeof(TPMT_HA)];
} TPM2B_DATA;

```

dgarske commented on src/tpm2.c:
```diff
@@ -5648,6 +5675,67 @@ UINT16 TPM2_GetVendorID(void)
+{
+    int rc;
+#ifndef WOLFTPM2_NO_WOLFCRYPT
+    int hashSize, nameAlgValue, nameAlgSize;
```
Please use `UINT16` for `nameAlgValue`.

tomoveu replied:
done. thanks

dgarske reviewed (changes requested):
Please correct feedback. Please add section in README.md with example output and usage. Please rebase and squash merge. The testing all passed and otherwise looks good.

tomoveu commented:
Awesome review @dgarske  Thank you 🥇 

All feedback is applied. Commits are squashed. Readme updated.

Please re-review and mark resolved issues :)

dgarske commented on examples/tpm_test.h:
```diff
@@ -43,6 +42,8 @@
 
 #define TPM2_DEMO_NV_TEST_INDEX         0x01800200
 #define TPM2_DEMO_NV_TEST_AUTH_INDEX    0x01800201
+#define TPM2_DEMO_NVRAM_STORE_PUB_INDEX 0x01800202
```
The pub index is not used... Please remove and rename `TPM2_DEMO_NVRAM_STORE_PRIV_INDEX`, since it contains both private and public portions.

tomoveu replied:
fixed. Thanks

dgarske commented on src/tpm2.c:
```diff
@@ -5648,6 +5665,69 @@ UINT16 TPM2_GetVendorID(void)
+    wc_HashAlg hash;
+    enum wc_HashType hashType;
+    byte appending[sizeof(TPMS_NV_PUBLIC)];
+    TPM2B_DATA digest;
```
This is an oversized buffer and will use more stack than required. Either use the correctly sized buffer `TPMU_NAME`, which is max digest + 2 bytes (handle) or use the incoming 'buffer` directly and remove this.

tomoveu replied:
I went with using directly buffer[2] , thanks.

dgarske commented on src/tpm2.c:
```diff
@@ -5648,6 +5665,69 @@ UINT16 TPM2_GetVendorID(void)
+    }
+
+    if (rc == 0) {
+        rc = wc_HashFinal(&hash, hashType, digest.buffer);
```
Put the hash digest directly into `&buffer[2]`.

tomoveu replied:
I was not sure if this is okay. I will do it like this using &buffer[2].

dgarske commented on src/tpm2.c:
```diff
@@ -5648,6 +5665,69 @@ UINT16 TPM2_GetVendorID(void)
+        nameAlgValue = TPM2_Packet_SwapU16(nvPublic->nameAlg);
+        nameAlgSize = sizeof(nvPublic->nameAlg);
+        XMEMCPY(buffer, (byte*)&nameAlgValue, nameAlgSize);
+        XMEMCPY(&buffer[2], digest.buffer, hashSize);
```
Then remove this.

tomoveu replied:
I removed the XMEMCPY regarding digest.buffer, the concatenation with nameAlg remains as required.

dgarske commented on src/tpm2_wrap.c:
```diff
@@ -2665,6 +2744,30 @@ int wolfTPM2_NVReadAuth(WOLFTPM2_DEV* dev, WOLFTPM2_NV* nv,
+    rc = wolfTPM2_SetNameHandle(dev, 0, &nv->handle);
+    rc |= wolfTPM2_SetNameHandle(dev, 1, &nv->handle);
+    if (rc != TPM_RC_SUCCESS) {
+        printf("Storing NV Index Name failed\n");
```
Please wrap this printf in `DEBUG_WOLFTPM`. The `rc` return code will not be valid with your above or'ing of the return code. Consider just returning `TPM_RC_FAILURE`.

tomoveu replied:
good point. Thank you

dgarske commented on src/tpm2_wrap.c:
```diff
@@ -2605,6 +2658,30 @@ int wolfTPM2_NVWriteAuth(WOLFTPM2_DEV* dev, WOLFTPM2_NV* nv,
+    rc = wolfTPM2_SetNameHandle(dev, 0, &nv->handle);
+    rc |= wolfTPM2_SetNameHandle(dev, 1, &nv->handle);
+    if (rc != TPM_RC_SUCCESS) {
+        printf("Storing NV Index Name failed\n");
```
Please wrap this printf in `DEBUG_WOLFTPM`. The `rc` return code will not be valid with your above or'ing of the return code. Consider just returning `TPM_RC_FAILURE`.

tomoveu replied:
Done. Thanks

dgarske commented on src/tpm2_wrap.c:
```diff
@@ -415,6 +430,24 @@ int wolfTPM2_SetAuthHandle(WOLFTPM2_DEV* dev, int index,
     return wolfTPM2_SetAuth(dev, index, TPM_RS_PW, auth, 0, name);
 }
 
+int wolfTPM2_SetNameHandle(WOLFTPM2_DEV* dev, int index,
```
Please rename this to `wolfTPM2_SetAuthHandleName`. It is just setting the name for an existing auth handle.

tomoveu replied:
OK. I think this name fits better, however the justification for me is different - We use the AUTH structure to pass the NAME of a TPM Handle down to our parameter encryption implementation.  Typically, the NAME of a TPM Handle is not part of a AUTH structure. In our case, it is a good solution to fetch necessary info for our parameter encryption.

dgarske commented on src/tpm2_wrap.c:
```diff
@@ -2560,9 +2602,19 @@ int wolfTPM2_NVCreateAuth(WOLFTPM2_DEV* dev, WOLFTPM2_HANDLE* parent,
+    }
+
     /* return new NV handle */
+    XMEMSET(nv, 0, sizeof(*nv));
```
This is broken. It will memset and clear the `nv->handle.name` done above. The memset needs moved above `TPM2_HashNvPublic`.

tomoveu replied:
Indeed, leftover after making the Hash Name computation a generic function. 

This actually had no impact over the examples, because we are making the Name computation here just in case some later command that will be using parameter encryption needs it.

Good catch!

tomoveu commented:
@dgarske  all feedback taken in, good points

dgarske commented on examples/README.md:
```diff
@@ -260,3 +268,29 @@ Loaded key to 0x80000001
+Private part = 222 bytes
+NV write succeeded
+
+$ ./examples/nvram/read 616 222 -aes
```
Having to supply the size of each public/private part seems problematic. Why not just used `sizeof(TPM2B_PUBLIC)` and `sizeof(TPM2B_PRIVATE)` like the `readKeyBlob` does? If you want to followup with this enhancement in PR #143 that is fine.

tomoveu replied:
Good question. As far as I know, TPM2B_PUBLIC is of size capable to take in the largest possible public key. Therefore, I thought having an argument to specify the size is a good approach. Similar for TPM2B_PRIVATE. I will check this and if I can avoid these manual arguments I will as addition to 143 Thanks for the hint!

tomoveu replied:
@dgarske coming back to this and related to #143 

I noticed that writeKeyBlob and readKeyBlob used in nvram/read use sizeof(key->pub) instead of the actual size of the Key Public Part. While this is okay for storing a file to disk, I am not sure if this is okay for using the NVRAM that on some TPMs is few kB only.

Example below:

In both cases, writeKeyBlob will write 616 bytes of public area, while the actual public area for the symmetric key is only 50 bytes and for RSA is only 278 bytes.

Should I just use sizeof(TPM2B_PUBLIC) for nvram/read or I should try to change the write/readKeyBlob helpers?

```
dimitartomov@Dimitars-MacBook-Pro wolfTPM % ./examples/keygen/keygen -sym
TPM2.0 Key generation example
	Key Blob: keyblob.bin
	Algorithm: SYMCIPHER
		AES, 256bits, CTR mode
	Template: Default
	Use Parameter Encryption: NULL
Loading SRK: Storage 0x81000200 (282 bytes)
AES Symmetric template with 256 bits, Counter mode
Creating new SYMCIPHER key...
Created new key (pub 50, priv 158 bytes)
Wrote 776 bytes to keyblob.bin
dimitartomov@Dimitars-MacBook-Pro wolfTPM % ./examples/keygen/keygen -t  
TPM2.0 Key generation example
	Key Blob: keyblob.bin
	Algorithm: RSA
	Template: Default
	Use Parameter Encryption: NULL
Loading SRK: Storage 0x81000200 (282 bytes)
RSA template
Creating new RSA key...
Created new key (pub 278, priv 222 bytes)
Wrote 840 bytes to keyblob.bin
dimitartomov@Dimitars-MacBook-Pro wolfTPM %
```

dgarske commented on examples/nvram/read.c:
```diff
@@ -0,0 +1,178 @@
+    readSize = keyBlob.priv.size;
+    printf("Trying to read %d bytes of private key part from NV\n", readSize);
+    rc = wolfTPM2_NVReadAuth(&dev, &nv, TPM2_DEMO_NVRAM_STORE_INDEX,
+        (byte*)&keyBlob.priv.buffer, &readSize, keyBlob.pub.size);
```
After loading the `keyBlob` can you add code to demonstrate loading the key?

tomoveu replied:
Good point. I will address that in 143 as well

dgarske commented on examples/README.md:
````diff
@@ -260,3 +268,29 @@ Loaded key to 0x80000001
+Typical output for storing and then reading an RSA key using parameter encryption:
+
+```
+$ ./examples/nvram/store -aes
````
Can we add support for specifying the key type? Or are you waiting for this to merge so you can do that as part of PR #143 ?

tomoveu replied:
Good idea, we could generate RSA or ECC. I would address this in 143 and we can have the parameter encryption fix merged here.

tomoveu replied:
@dgarske RSA ECC or Symmetric key all work with nvram/store,

because nvram/store uses readKeyBlob

and whatever is written in the KeyBlob->pub/priv.size that is then stored into the NVRAM (size-wise).

Did you mean to be able to store only the private material (-priv option perhaps)?

Could be useful for symmetric keys. Also -pub in cases when only the pub part has to be stored in NVRAM.

dgarske commented on examples/nvram/store.c:
```diff
@@ -0,0 +1,170 @@
+    WOLFTPM2_NV nv;
+    TPM2B_AUTH auth;
+    word32 nvAttributes;
+    const char* filename = "keyblob.bin";
```
Can we make this a command line argument? Or followup with another PR?

tomoveu replied:
Let's follow-up with another PR for NVRAM examples enhancements

tomoveu replied:
nvram/store already accepts filename, but nvram/read does not. I will make sure to add that.

dgarske reviewed (changes requested):
Looks good, but there are some build errors with `--disable-wolfcrypt`:

```
examples/tpm_test_keys.c:256:33: error: unused parameter 'tpmDevId' [-Werror,-Wunused-parameter]
                            int tpmDevId,
                                ^
examples/tpm_test_keys.c:290:33: error: unused parameter 'tpmDevId' [-Werror,-Wunused-parameter]
                            int tpmDevId,
                                ^
2 errors generated.

src/tpm2.c:109:20: error: unused variable 'handleValue' [-Werror,-Wunused-variable]
    UINT32 authSz, handleValue;
                   ^
src/tpm2.c:112:21: error: unused variable 'handlePos' [-Werror,-Wunused-variable]
    int i, authPos, handlePos;
                    ^
src/tpm2.c:5671:9: error: unused variable 'rc' [-Werror,-Wunused-variable]
    int rc;
        ^
src/tpm2.c:5669:39: error: unused parameter 'nvPublic' [-Werror,-Wunused-parameter]
int TPM2_HashNvPublic(TPMS_NV_PUBLIC* nvPublic, byte* buffer, UINT16* size)
                                      ^
src/tpm2.c:5669:55: error: unused parameter 'buffer' [-Werror,-Wunused-parameter]
int TPM2_HashNvPublic(TPMS_NV_PUBLIC* nvPublic, byte* buffer, UINT16* size)
                                                      ^
src/tpm2.c:5669:71: error: unused parameter 'size' [-Werror,-Wunused-parameter]
int TPM2_HashNvPublic(TPMS_NV_PUBLIC* nvPublic, byte* buffer, UINT16* size)
                                                                      ^
6 errors generated.
```

tomoveu commented:
@dgarske good recommendations for enhancements on the NVRAM examples, I will do them in follow-up commit.

I added fixes for when wolfCrypt is disabled. Thank you for catching this 💯
